PURL-TYPES.rst

```diff
@@ -521,6 +521,44 @@ swift
       pkg:swift/github.com/Alamofire/Alamofire@5.4.3
       pkg:swift/github.com/RxSwiftCommunity/RxFlow@2.12.4
 
+vcpkg
+-----
+``vcpkg`` for `vcpkg <https://vcpkg.io>`_ C and C++ packages:
+
+- By default, vcpkg will build C and C++ dependencies from source, using configurations compatible with the user's project. For this and other reasons, it uses different terminology, documented here.
+
+  - **port** - A package, along with it's build scripts and possibly minor modifications.
+  - **registry** - A collection of ports, possibly private to the user. Analogous to **repository**.
+
+- ``namespace``: Currently reserved for future use and should be empty
+
+- ``name``: The case-sensitive name of the port. **Required**
+- ``version``: The port version. **Required**
+- ``qualifiers``: The qualifiers below are used to provide more specific information on a port's origin registry and to distinguish between multiple instances of the same port dependency within the same project (for example, when the project targets multiple platforms).
+
+  - ``repository_url``: A `percent-encoded <https://www.rfc-editor.org/rfc/rfc3986#section-2.1>`_ absolute URI for the registry the port came from. For filesystem registries, the URI will have a `file` URI scheme. **Optional**
+
+    - If empty, the port comes from the default registry, ``https://github.com/microsoft/vcpkg``.
+
+  - ``registry_version``: A string specifying the version of the registry specified by `repository_url`. 
+  
+    - For git registries, the commit hash (potentially abbreviated). This must __NOT__ be any other kind of git reference, such as `HEAD` or tags. **Required**
+    - For filesystem registries, an string that can be used to identify specific versions, such as Coordinated Universal Time (UTC) as described in `RFC 3339 <https://datatracker.ietf.org/doc/html/rfc3339>`_. If using a UTC time, be sure to percent-encode it correctly. **Optional**
+
+  - Additional qualifiers may be present on the purl that provide additional information about the context in which the package is being used, such as build configuration or platform information. These additional qualifiers must be tolerated during parsing and can be ignored if the parser does not expect them.
+
+- ``subpath``: Currently reserved for future use and should be empty.
+- Examples::
+
+    Microsoft GSL version 4.0.0 from the default registry at commit 4e2b371
+        pkg:vcpkg/ms-gsl@4.0.0?registry_version=4e2b371 
+    ffmpeg version 5.1.2 from the azure-sdk registry at commit e6c17cc
+        pkg:vcpkg/ffmpeg@5.1.2?registry_url=https%3A%2F%2Fgithub.com%2Fazure-sdk%2Fvcpkg&registry_version=e6c17cc
+    LLVM from a filesystem registry
+        pkg:vcpkg/llvm@15.0.7?registry_url=file%3A%2F%2F%2FC%3A%2Flocal-registry%2Fvcpkg&registry_version=2023-08-18T22%3A28%3A52Z
```

Comment:
These are over escaped. It's valid to write them that way and parsers need to be prepared to accept it, but it's not the canonical format and they probably shouldn't be written this way in the spec examples because it propagates confusion about how escaping is intended to be performed (see comments on #39).

Compliant parsers should handle `pkg:vcpkg/llvm@15.0.7?registry_url=file:///C:/local-registry/vcpkg&registry_version=2023-08-18T22:28:52Z` exactly the same as `pkg:vcpkg/llvm@15.0.7?registry_url=file%3A%2F%2F%2FC%3A%2Flocal-registry%2Fvcpkg&registry_version=2023-08-18T22%3A28%3A52Z`. These qualifier values _are_ percent encoded, but none of the characters require replacement during encoding according to WHATWG URL query string or PURL qualifier rules¹, so the encoded and unencoded strings are the same. If the `registry_version` qualifier value was `2023-08-18 22:28:52Z` then it would need to be encoded as `2023-08-18%2022:28:52Z`. Maybe if you want to demonstrate the escaping the file path contains a space.

Breaking it down following the PURL parsing algorithm defined in the spec:
- Split on `?`: `pkg:vcpkg/llvm@15.0.7` `registry_url=file:///C:/local-registry/vcpkg&registry_version=2023-08-18T22:28:52Z`
- Split on `&`: `pkg:vcpkg/llvm@15.0.7` `registry_url=file:///C:/local-registry/vcpkg` `registry_version=2023-08-18T22:28:52Z`
- Split on `=`: `pkg:vcpkg/llvm@15.0.7` `registry_url` `file:///C:/local-registry/vcpkg` `registry_version` `2023-08-18T22:28:52Z`
- Percent decode qualifiers: `pkg:vcpkg/llvm@15.0.7` `registry_url` `file:///C:/local-registry/vcpkg` `registry_version` `2023-08-18T22:28:52Z`

¹ For PURL, I think the qualifier encoding rules are best defined as WHATWG URL [query percent-encode set](https://url.spec.whatwg.org/#query-percent-encode-set) and `@`, `?`, and `#` (["must be encoded as %[…] elsewhere"](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#character-encoding)). I think these additions by the PURL spec are to simplify parsing (PURL finds `?` from the right early on but WHATWG URL finds `?` from the left while processing the path component). PURL specifically states that `:` and `/` never require encoding.

Reply:
The comments on #39, as I understood them, reinforce that these are properly percent-encoded; despite your most recent comment proposing otherwise.

You may think that the qualifier encoding rules are best defined as WHATWG URL query percent-encode set, but that's not what the purl specification says. I would agree that the step-by-step parsing that is described should handle not encoding these reserved characters, but those instructions feel non-normative (meant to provide a concrete example of how it *may* be done).

It's fairly clear from several parts of the specification that the `value` component of a `key=value` qualifier should be UTF-8 encoded and then percent encoded. Percent encoding is explicitly defined as that at the Wikipedia page, which references RFC 3986, which is unambiguous that all reserved characters should be percent encoded. That is to say, not the WHATWG URL encoding rules.

Specifically, the section on qualifiers states

```
A value must be a percent-encoded string
```

There doesn't seem to be much ambiguity there. There is a line in the encoding section that mention certain characters should not be encoded when they appear as separators, but it's clear from the context that they mean separators of the component parts of the purl.

Furthermore, there is a example of this in the README.rst file here it shows `pkg:maven/org.apache.xmlgraphics/batik-anim@1.9.1?repository_url=repo.spring.io%2Frelease` which has an encoded `/` in the qualifier value.

Ultimately, I think that certain, incorrectly encoded qualifier values should be accepted by parsers, since it is possible to implement a parser that can tolerate them (but only certain cases; for instance you would need to encode `?` if it appeared). But it seems that the canonical representation should be the correctly percent-encoded values, not the other way around.

Reply:
Percent encoding is just a method of encoding. It does not specify which characters must be encoded. RFC 3986 section 2 says

> If data for a URI component would conflict with a reserved character's purpose as a delimiter, then the conflicting data must be percent-encoded before the URI is formed.
> […]
> A subset of the reserved characters (gen-delims) is used as delimiters of the generic URI components described in [Section 3](https://www.rfc-editor.org/rfc/rfc3986#section-3).  A component's ABNF syntax rule will not use the reserved or gen-delims rule names directly; instead, each syntax rule lists the characters allowed within that component (i.e., not delimiting it), and any of those characters that are also in the reserved set are "reserved" for use as subcomponent delimiters within the component.
> […]
> URI producing applications should percent-encode data octets that correspond to characters in the reserved set unless these characters are specifically allowed by the URI scheme to represent data in that component.

So the reserved set is not the whole story when it comes to URI construction.

Section 3 defines that paths are composed from runs of `pchar`, which is allowed to contain unreserved characters and also `:` and `@` (used by PURL) and any `sub-delim`. https://www.rfc-editor.org/rfc/rfc3986#section-3.3

Then it says that queries can contain any `pchar` and also slashes and question marks.

So if PURL removes `@` `?` `#` from the allowable character set, that leaves us with the following allowed characters:

unreserved / pct-encoded / sub-delims / `:` / `/`

Or, in the format used by WHATWG, the characters that must be encoded are:

"unreserved" (in this case listing the reserved): C0 controls, ` ` `!` `"` `#` `$` `&` `'` `(` `)` `*` `+` `,` `/` `:` `;` `<` `=` `>` `?` `@` `[` `\` `]` `^` <code>\`</code> `{` `|` `}`  
"sub-delim": C0 controls, ` ` `"` `#` `/` `:` `<` `>` `?` `@` `[` `\` `]` `^` <code>\`</code> `{` `|` `}`  
"pchar": C0 controls, ` ` `"` `#` `/` `<` `>` `?` `[` `\` `]` `^` <code>\`</code> `{` `|` `}`  
"query": C0 controls, ` ` `"` `#` `<` `>` `[` `\` `]` `^` <code>\`</code> `{` `|` `}`

And then disallowing `@` `?` `#` from PURL, we arrive at

qualifiers: C0 controls, ` ` `"` `#` `<` `>` `?` `@` `[` `\` `]` `^` <code>\`</code> `{` `|` `}`

Which can also be written:

qualifiers: WHATWG URL query percent-encode set, `?` `@` `[` `\` `]` `^` <code>\`</code> `{` `|` `}`

RFC 3986 URI-based PURL requires more characters be encoded than WHATWG URL-based PURL, but it does not require the slashes or colons we're concerned about to be encoded.

Most of these additional required-to-be-encoded characters in RFC 3986 have no meaning anywhere in a URI so even if unencoded they shouldn't affect parsing. `[` and `]` have special meaning in the authority, but not in a way that those characters could be ambiguous when used anywhere else in a valid URI.

WHATWG URL is intended to replace RFC3986 URI.

Reply:
I'm not convinced by your argument, but I'd be happy to change the input to the test if one of the @package-url maintainers agrees with you. Not having to percent-encode those characters in a qualifier value does seem simpler, but the spec currently does not reference the WHATWG URL rules.

Reply:
This has been changed now to be consistent with the other examples.